### PR TITLE
Initial commit with action github action code

### PR DESCRIPTION
**Pull request by Luke-Roy-IBM**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,154 @@
...
+
+This GitHub Action allows you to interact with IBM Cloud Code Engine. Deploy Apps, Jobs, and Functions. It offers flexibility for different deployment types and provides various configuration options.
+
+## Inputs
+
+| Name            | Required | Default Value |Description |
+|-----------------|----------|---------------|----------------------------------------------------------------|
+| `api-key`       | ✅      | -             | IAM API Key used to log into the IBM Cloud. Please store your IBM Cloud API key securely in your GitHub repository Secrets.|
+| `resource-group`| ❌       | Default       | An IBM Cloud Resource Group, a logical container for organizing and managing related cloud resources.|
+| `region`        | ✅      | -             | The geographical area where your Code Engine project is located.|
```
Comment (reggeenr) [inline README.md]:
```suggestion
| `region`        | ✅      | -             | The geographical area where your Code Engine project is located, like `eu-de` |
```

File: README.md [hunk 2]
```diff
@@ -1,2 +1,154 @@
...
+| `project`       | ✅      | -             | A Code Engine Project, grouping your Apps, Functions, and Jobs.|
```
Comment (reggeenr) [inline README.md]:
```suggestion
| `project`       | ✅      | -             | The unique identifier (GUID) or the name that identifies your IBM Cloud Code Engine project. |
```

File: README.md [hunk 3]
```diff
@@ -1,2 +1,154 @@
...
+| `entity`        | ✅      | -             | The type of entity to deploy (App, Function, Job). |
```
Comment (reggeenr) [inline README.md]:
```suggestion
| `component`        | ✅      | -             | The type of component to deploy. allowed values `app`, `function`, `job` |
```

File: action.yml [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    - name: Install Code Engine Plugin
+      shell: bash
+      run: ibmcloud plugin install code-engine
+
+    # Select the project using Name or ID
+    # If the project doesn`t exist the project is created with the provided Name
+    - name: Select Code Engine Project
+      shell: bash
+      run: | 
+        if ibmcloud ce project select --name ${{ inputs.project }} || ibmcloud ce project select --id ${{ inputs.project }} ; then
```
Comment (reggeenr) [inline action.yml]: encapsulate the name with quotes to enable like 'My project'

File: README.md [hunk 5]
```diff
@@ -1,2 +1,154 @@
...
+| `name`          | ✅      | -             | The name of the App, Function, or Job.|
+| `runtime`       | ❌ | -             | The runtime used for the Function. Currently supported `nodejs-18` and `python-3.11` see [IBM Code Engine Function Runtimes](https://cloud.ibm.com/docs/codeengine?topic=codeengine-fun-runtime) for more information.|
+| `build-source`  | ❌       | .             | Path to the directory containing the source code.|
+| `cpu`           | ❌       | -             | CPU value set for your entity [Config for Functions](https://cloud.ibm.com/docs/codeengine?topic=codeengine-fun-runtime), [Codeengine Memory CPU combo](https://cloud.ibm.com/docs/codeengine?topic=codeengine-mem-cpu-combo)|
```
Comment (reggeenr) [inline README.md]: link must point to cpu mem values that are allowed: https://cloud.ibm.com/docs/codeengine?topic=codeengine-mem-cpu-combo

**Review by reggeenr — COMMENTED** [review]
Great job, @Luke-Roy-IBM!

Please see my comments. I will finish my review as soon as possible

File: README.md (hunk 5 above)
Comment (Luke-Roy-IBM) [reply README.md]: Already in the description

File: action.yml (hunk 4 above)
Comment (Luke-Roy-IBM) [reply action.yml]: fixed

File: action.yml [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+      run: |
+        ibmcloud login --apikey ${{ inputs.api-key }} -r ${{ inputs.region }}
+        
+    - name: Set Resource Group
+      shell: bash
+      run: |
+        if [[ "${{ inputs.resource-group }}" != "" ]] ; then
+          ibmcloud target -g ${{ inputs.resource-group }}
+        else
+          ibmcloud target -g $(ibmcloud resource groups --default --output JSON | jq -r '.[0].id')
```
Comment (reggeenr) [inline action.yml:73]: I like!

File: action.yml [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+  region:
+    description: The Region refers to a geographical area where the project is are located
+    required: true
+
+  project:
+    description: A Code Engine Project Is the grouping of your Apps, Functions and Jobs
+    required: true
+
+# App, Function or Job Specific inputs
+  entity:
```
Comment (reggeenr) [inline action.yml]: We should rename `entity` to `component`

File: action.yml [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+          echo "memory=--memory ${{ inputs.memory }}" >> "$GITHUB_OUTPUT"
+        else
+          echo "memory=" >> "$GITHUB_OUTPUT"
+        fi
+
+    # Functions Steps
+    - name: Create or Update Functions
+      shell: bash
+      id: fn-create
+      if: ( inputs.entity == 'function' || inputs.entity == 'func' || inputs.entity == 'fn' )
```
Comment (reggeenr) [inline action.yml]:
```suggestion
      if: ( inputs.component == 'function' || inputs.component == 'func' || inputs.component == 'fn' ) 
```

File: action.yml [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+          ibmcloud ce fn update --name ${{ inputs.name }} --runtime ${{ inputs.runtime }} --build-source ${{ inputs.build-source }} ${{ steps.set-resources.outputs.cpu }} ${{ steps.set-resources.outputs.memory }}
+        else
+          ibmcloud ce fn create --name ${{ inputs.name }} --runtime ${{ inputs.runtime }} --build-source ${{ inputs.build-source }} ${{ steps.set-resources.outputs.cpu }} ${{ steps.set-resources.outputs.memory }}
+        fi
+
+    # Application Steps
+    - name: Create or Update Application
+      shell: bash
+      id: create-app
+      if: ( inputs.entity == 'application' || inputs.entity == 'app' )
```
Comment (reggeenr) [inline action.yml]:
```suggestion
      if: ( inputs.component == 'application' || inputs.component == 'app' ) 
```

File: action.yml [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+          ibmcloud ce application update --name ${{ inputs.name }} --build-source ${{ inputs.build-source }} ${{ steps.set-resources.outputs.cpu }} ${{ steps.set-resources.outputs.memory }}
+        else
+          ibmcloud ce application create --name ${{ inputs.name }} --build-source ${{ inputs.build-source }} ${{ steps.set-resources.outputs.cpu }} ${{ steps.set-resources.outputs.memory }}
+        fi
+
+    # Job Steps
+    - name: Create or Update Job
+      shell: bash
+      id: create-job
+      if: inputs.entity == 'job'
```
Comment (reggeenr) [inline action.yml]:
```suggestion
      if: inputs.component == 'job' 
```

File: action.yml [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+          ibmcloud ce job update --name ${{ inputs.name }} --build-source ${{ inputs.build-source }} --wait ${{ steps.set-resources.outputs.cpu }} ${{ steps.set-resources.outputs.memory }}
+        else
+          ibmcloud ce job create --name ${{ inputs.name }} --build-source ${{ inputs.build-source }} --wait ${{ steps.set-resources.outputs.cpu }} ${{ steps.set-resources.outputs.memory }}
+        fi
+    
+    - name: Get entity 
+      shell: bash
+      if: steps.fn-create.outcome == 'success' || steps.app-create.outcome == 'success' || steps.job-create.outcome == 'success'
+      run: |
+        case ${{ inputs.entity }} in
```
Comment (reggeenr) [inline action.yml]:
```suggestion
        case ${{ inputs.component }} in
```

File: action.yml [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+    description: Runtime used for the Function only required for function
+    required: false
+
+  build-source:
+    description: path to the directory containing the source code 
+    required: false
+    default: .
+
+  cpu:
+    description: CPU configuration set for the entity. If not set default Code Engine values are used.
```
Comment (reggeenr) [inline action.yml]:
```suggestion
    description: CPU configuration set for the component. If not set default Code Engine values are used. 
```

File: action.yml [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+    description: path to the directory containing the source code 
+    required: false
+    default: .
+
+  cpu:
+    description: CPU configuration set for the entity. If not set default Code Engine values are used. 
+    required: false
+
+  memory:
+    description: Memory configuration set for the entity. If not set default Code Engine values are used.
```
Comment (reggeenr) [inline action.yml]:
```suggestion
    description: Memory configuration set for the component. If not set default Code Engine values are used.
```

File: action.yml [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+      if: inputs.entity == 'job' 
+      run: |
+
+        if ibmcloud ce job get --name ${{ inputs.name }} ; then
+          ibmcloud ce job update --name ${{ inputs.name }} --build-source ${{ inputs.build-source }} --wait ${{ steps.set-resources.outputs.cpu }} ${{ steps.set-resources.outputs.memory }}
+        else
+          ibmcloud ce job create --name ${{ inputs.name }} --build-source ${{ inputs.build-source }} --wait ${{ steps.set-resources.outputs.cpu }} ${{ steps.set-resources.outputs.memory }}
+        fi
+    
+    - name: Get entity
```
Comment (reggeenr) [inline action.yml]:
```suggestion
    - name: Get component 
```

File: action.yml [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+                ibmcloud ce fn get --name ${{ inputs.name }}
+                ;;
+            application|app)
+                ibmcloud ce app get --name ${{ inputs.name }}
+                ;;
+            job)
+                ibmcloud ce job get --name ${{ inputs.name }}
+                ;;
+            *)
+                echo "Wrong Code Engine Entity used!"
```
Comment (reggeenr) [inline action.yml]:
```suggestion
                echo "Wrong Code Engine component used!"
```

File: action.yml [hunk 16]
```diff
@@ -0,0 +1,168 @@
...
+
+    # Select the project using Name or ID
+    # If the project doesn`t exist the project is created with the provided Name
+    - name: Select Code Engine Project
+      shell: bash
+      run: | 
+        if ibmcloud ce project select --name "${{ inputs.project }}" || ibmcloud ce project select --id ${{ inputs.project }} ; then
+          echo "Project Selected"
+        else
+          ibmcloud ce project create --name ${{ inputs.project }}
```
Comment (reggeenr) [inline action.yml]:
```suggestion
          ibmcloud ce project create --name "${{ inputs.project }}"
```

**Review by reggeenr — CHANGES_REQUESTED** [review]
Hi @Luke-Roy-IBM , just added a few suggestions on renaming.

File: action.yml (hunk 7 above)
Comment (Luke-Roy-IBM) [reply action.yml]: Added the changes in the readme

**Review by reggeenr — APPROVED** [review]
LGTM - great stuff. can't wait to see it in action

**Pull request merged**